### PR TITLE
gzip: Handle encode trailers for gzip http filter

### LINKED ISSUE #6894: GZip filter + grpc_json_transcoder doesn't work on some client
GZip filter + grpc_json_transcoder doesn't work on some client

*Description*:
We have noticed that some clients are reporting invalid gzip streams for requests done through gzip+grpc_json_transcoder filters. Other requests simply going through the gzip filter do not have this issue.

For example, given the following config:

```
admin:
  access_log_path: /tmp/admin_access.log
  address: { socket_address: { address: 0.0.0.0, port_value: 8900 } }

static_resources:
  listeners:
  - name: main_listener
    address: { socket_address: { address: 0.0.0.0, port_value: 8000 } }
    filter_chains:
    - filters:
      - name: envoy.http_connection_manager
        config:
          stat_prefix: ingress_http
          codec_type: auto
          use_remote_address: true
          route_config:
            name: local_route
            virtual_hosts:
            - name: backend
              domains: ["*"]
              routes:
              - match: { prefix: "/plain" }
                route: { cluster: plain_hello }
              - match: { prefix: "/" }
                route: { cluster: grpc_hello }
          http_filters:
          - name: envoy.gzip
          - name: envoy.grpc_json_transcoder
            config:
              proto_descriptor: /app/descriptors.pb
              services: ["HelloService.HelloService"]
          - name: envoy.router
    perConnectionBufferLimitBytes: 16384000

  clusters:
    - name: grpc_hello
      connect_timeout: 1.0s
      type: logical_dns
      lb_policy: round_robin
      http2_protocol_options: {}
      hosts:
      - socket_address: { address: 127.0.0.1, port_value: 30000 }
    - name: plain_hello
      connect_timeout: 1.0s
      type: logical_dns
      lb_policy: round_robin
      hosts:
      - socket_address: { address: 127.0.0.1, port_value: 5000 }
```

A request to the `plain_hello` endpoint with gzip compression works fine:
```
$ curl -H "Accept-Encoding: gzip" -v http://127.0.0.1:8000/plain/hello | gunzip
*   Trying 127.0.0.1...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /plain/hello HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.47.0
> Accept: */*
> Accept-Encoding: gzip
> 
< HTTP/1.1 200 OK
< content-type: text/html; charset=utf-8
< server: envoy
< date: Fri, 10 May 2019 16:54:16 GMT
< x-envoy-upstream-service-time: 1
< vary: Accept-Encoding
< content-encoding: gzip
< transfer-encoding: chunked
< 
{ [74 bytes data]
100    58    0    58    0     0  11225      0 --:--:-- --:--:-- --:--:-- 14500
* Connection #0 to host 127.0.0.1 left intact
plain hello world! 012345678901234567890123456789
```

While a request to the `grpc_hello` endpoint with gzip compression produces an error:
```
$ curl -H "Accept-Encoding: gzip" -v http://127.0.0.1:8000/grpc/hello | gunzip
*   Trying 127.0.0.1...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /grpc/hello HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.47.0
> Accept: */*
> Accept-Encoding: gzip
> 
< HTTP/1.1 200 OK
< content-type: application/json
< grpc-accept-encoding: identity,deflate,gzip
< accept-encoding: identity,gzip
< x-envoy-upstream-service-time: 0
< grpc-status: 0
< grpc-message: 
< vary: Accept-Encoding
< content-encoding: gzip
< date: Fri, 10 May 2019 16:55:58 GMT
< server: envoy
< transfer-encoding: chunked
< 
{ [68 bytes data]
100    57    0    57    0     0  37279      0 --:--:-- --:--:-- --:--:-- 57000
* Connection #0 to host 127.0.0.1 left intact
{"text":"gRPC hello world! 012345678901234567890123456789"}
gzip: stdin: unexpected end of file
```

Note that the payload is ok, it is just the gzip output that appears to be missing some trailing bytes. 

This looks similar to what was reported in #2909 and fixed in #3025 .  This occurs on all Envoy versions I have tested (v1.7.0, v1.8.0, v1.10.0), (with the particularity that I need to change where the gzip filter is inserted in older Envoy configs).

Here are the Envoy v1.10.0 log when producing a  "bad" gzip stream:
```
[2019-05-10 16:57:48.375][25][debug][main] [source/server/connection_handler_impl.cc:257] [C4] new connection
[2019-05-10 16:57:48.375][25][debug][http] [source/common/http/conn_manager_impl.cc:243] [C4] new stream
[2019-05-10 16:57:48.375][25][debug][http] [source/common/http/conn_manager_impl.cc:580] [C4][S10282197792249075826] request headers complete (end_stream=true):
':authority', '127.0.0.1:8000'
':path', '/grpc/hello'
':method', 'GET'
'user-agent', 'curl/7.47.0'
'accept', '*/*'
'accept-encoding', 'gzip'

[2019-05-10 16:57:48.375][25][debug][http] [source/common/http/conn_manager_impl.cc:1037] [C4][S10282197792249075826] request end stream
[2019-05-10 16:57:48.375][25][debug][router] [source/common/router/router.cc:320] [C4][S10282197792249075826] cluster 'grpc_hello' match for URL '/HelloService.HelloService/SayHello'
[2019-05-10 16:57:48.375][25][debug][router] [source/common/router/router.cc:381] [C4][S10282197792249075826] router decoding headers:
':authority', '127.0.0.1:8000'
':path', '/HelloService.HelloService/SayHello'
':method', 'POST'
':scheme', 'http'
'user-agent', 'curl/7.47.0'
'accept', '*/*'
'accept-encoding', 'gzip'
'x-forwarded-for', '192.168.56.202'
'x-forwarded-proto', 'http'
'x-envoy-internal', 'true'
'x-request-id', 'b809a597-c075-4096-b519-127074ce1b84'
'content-type', 'application/grpc'
'x-envoy-original-path', '/grpc/hello'
'te', 'trailers'
'x-envoy-expected-rq-timeout-ms', '15000'

[2019-05-10 16:57:48.375][25][debug][pool] [source/common/http/http2/conn_pool.cc:96] [C3] creating stream
[2019-05-10 16:57:48.375][25][debug][router] [source/common/router/router.cc:1165] [C4][S10282197792249075826] pool ready
[2019-05-10 16:57:48.376][25][debug][router] [source/common/router/router.cc:717] [C4][S10282197792249075826] upstream headers complete: end_stream=false
[2019-05-10 16:57:48.376][25][debug][client] [source/common/http/codec_client.cc:95] [C3] response complete
[2019-05-10 16:57:48.376][25][debug][pool] [source/common/http/http2/conn_pool.cc:237] [C3] destroying stream: 0 remaining
[2019-05-10 16:57:48.376][25][debug][http] [source/common/http/conn_manager_impl.cc:1278] [C4][S10282197792249075826] encoding headers via codec (end_stream=false):
':status', '200'
'content-type', 'application/json'
'grpc-accept-encoding', 'identity,deflate,gzip'
'accept-encoding', 'identity,gzip'
'x-envoy-upstream-service-time', '0'
'grpc-status', '0'
'grpc-message', ''
'vary', 'Accept-Encoding'
'content-encoding', 'gzip'
'date', 'Fri, 10 May 2019 16:57:48 GMT'
'server', 'envoy'

[2019-05-10 16:57:48.376][25][debug][http] [source/common/http/conn_manager_impl.cc:1446] [C4][S10282197792249075826] encoding trailers via codec:
'grpc-status', '0'
'grpc-message', ''

[2019-05-10 16:57:48.376][25][debug][http2] [source/common/http/http2/codec_impl.cc:577] [C3] stream closed: 0
[2019-05-10 16:57:48.376][25][debug][connection] [source/common/network/connection_impl.cc:502] [C4] remote close
[2019-05-10 16:57:48.376][25][debug][connection] [source/common/network/connection_impl.cc:183] [C4] closing socket: 0
[2019-05-10 16:57:48.376][25][debug][main] [source/server/connection_handler_impl.cc:68] [C4] adding to cleanup list
```

I've attached a sample minimal python application + envoy config that reproduce the issue (which is that I used to produce the above traces).

[envoy_gzip_issue.tar.gz](https://github.com/envoyproxy/envoy/files/3167453/envoy_gzip_issue.tar.gz)

### PR DESCRIPTION
**Pull request by dio**
Description: Compressor's compress needs to be called with Z_FINISH when end stream is true or at the encode trailers handler. 

Risk Level: Low
Testing: Unit test is provided
Docs Changes: N/A
Release Notes: N/A
Fixes #6894 

Signed-off-by: Dhi Aurrahman <dio@tetrate.io>

### CODE REVIEW

**Comment by dio** [comment]
@gsagula @lizan can you help me to review this? Thanks!

**Comment by dio** [comment]
@alexandrecp-b can you help to verify the fix? Thanks!

File: source/extensions/filters/http/gzip/gzip_filter.cc [hunk 1]
```diff
@@ -146,6 +146,16 @@ Http::FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_s
   return Http::FilterDataStatus::Continue;
 }
 
+Http::FilterTrailersStatus GzipFilter::encodeTrailers(Http::HeaderMap&) {
```
Comment (lizan) [inline source/extensions/filters/http/gzip/gzip_filter.cc]: can we do similar thing in `decodeTrailers` path as well?

Comment (lizan) [reply source/extensions/filters/http/gzip/gzip_filter.cc]: ah nvm I thought we do decompress in decode path but that's not the case.

**Comment by alexandrecp-b** [comment]
@dio : I confirm that this fix does resolve #6894 for me. Thanks!

**Pull request merged**